### PR TITLE
[serve] Drop stale handle metrics 

### LINKED ISSUE #43998: [serve] autoscaling metrics collected from handle is not cleared correctly
### What happened + What you expected to happen

I've observed in some cases that even after there are no more requests being sent, the metrics from handles that's being used by the controller to make autoscaling decisions stays at a nonzero number, and the inaccurate number means the metrics is not being cleared somewhere. Some possible reasons:
* Race condition when incrementing/decrementing requests
  * This possibility can probably be eliminated because we have a threading lock over every place we modify the counter.
* We cannot always rely on object ref `._on_completed()` being called?
  * Not sure if this is true, need to verify.
* Nonzero handle metrics get pushed to controller, handle is removed, and the nonzero handle metrics live in the controller forever. This is something that can for sure happen, but unclear if this is the culprit.

### Versions / Dependencies

ray 2.10

### Reproduction script

n/a, unable to come up with repro

### Issue Severity

None

### LINKED ISSUE #44046: [serve] Handle metrics are not cleaned up on handle exit, preventing downscaling
We don't currently drop handle metrics after a handle has been removed (i.e., a replica or proxy with a handle has been shut down). This causes "phantom" metrics to prevent downscaling.

This only affects correctness if the handle had ongoing requests when it was removed (hence why it wasn't caught in our manual testing).

This is introduced by the new `RAY_SERVE_COLLECT_AUTOSCALING_METRICS_ON_HANDLE` change (which is being turned on by default in 2.10).

### PR DESCRIPTION
**Pull request by edoakes**
## Why are these changes needed?

We don't currently drop handle metrics after a handle has been removed (i.e., a replica or proxy with a handle has been shut down). This causes "phantom" metrics to prevent downscaling.

This PR adds a timeout after which we drop the metrics: 2 * the reporting period (with a constant floor of 10s in case the reporting period is configured to be very low).

Also cleans up a few log statements.

## Related issue number

Closes https://github.com/ray-project/ray/issues/44046
Closes https://github.com/ray-project/ray/issues/43998

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/_private/deployment_state.py [hunk 1]
```diff
@@ -1608,6 +1609,27 @@ def get_total_num_requests(self) -> float:
             RAY_SERVE_COLLECT_AUTOSCALING_METRICS_ON_HANDLE
             or len(running_replicas) == 0
         ):
+            # Remove metrics for handles that haven't sent updates in a while.
+            # Clear metrics for handles that haven't sent an update in awhile.
```
Comment (zcin) [inline python/ray/serve/_private/deployment_state.py:1613]: Remove one of these lines

**Review by zcin — APPROVED** [review]

**Comment by edoakes** [comment]
Merged in https://github.com/ray-project/ray/pull/44049 so CI would pass, codeowners please ignore

**Review by shrekris-anyscale — COMMENTED** [review]
I reviewed the code and the tests. The change looks good overall.

File: python/ray/serve/_private/deployment_state.py [hunk 2]
```diff
@@ -1608,6 +1609,26 @@ def get_total_num_requests(self) -> float:
...
+                    del self.handle_requests[handle_id]
+                    if (
+                        handle_metric.queued_requests
+                        + sum(handle_metric.running_requests.values())
+                        > 0
+                    ):
+                        logger.info(
+                            f"Dropping stale metrics for handle '{handle_id}' "
+                            f"(no updates received for {timeout_s:.1f}s)."
+                        )
```
Comment (shrekris-anyscale) [inline python/ray/serve/_private/deployment_state.py:1630]: Can you log the number of queued requests at the handle here? If this causes odd autoscaling behavior, that would help us figure out how much of an impact this staleness-handling has.

File: python/ray/serve/_private/deployment_state.py [hunk 3]
```diff
@@ -1608,6 +1609,26 @@ def get_total_num_requests(self) -> float:
             RAY_SERVE_COLLECT_AUTOSCALING_METRICS_ON_HANDLE
             or len(running_replicas) == 0
         ):
+            # Drop metrics for handles that haven't sent an update in awhile.
```
Comment (shrekris-anyscale) [inline python/ray/serve/_private/deployment_state.py:1612]:
```suggestion
            # Drop metrics for handles that haven't sent an update in a while.
```

File: python/ray/serve/_private/autoscaling_policy.py [hunk 4]
```diff
@@ -110,3 +110,7 @@ def apply_bounds(
             target_capacity_direction,
         )
         return max(lower_bound, min(upper_bound, curr_target_num_replicas))
+
+    def get_metrics_interval_s(self) -> Optional[float]:
+        if self.config:
+            return self.config.metrics_interval_s
```
[the remark this answers is not in the document]
Comment (zcin) [reply python/ray/serve/_private/autoscaling_policy.py:116]: Yeah this should only get called if the autoscaling config is non-null, and autoscaling config is a pydantic model so it makes sure the `metrics_interval_s` is a number at runtime.

File: python/ray/serve/_private/deployment_state.py [hunk 5]
```diff
@@ -1662,13 +1678,25 @@ def autoscale(self) -> int:
...
 
+        curr_stats_str = (
+            f"Current ongoing requests: {total_num_requests:.2f}, "
+            f"current running replicas: {num_running_replicas}."
+        )
         new_num = self._target_state.target_num_replicas
         if new_num > old_num:
+            logger.info(
+                f"Upscaling {self._id} from {old_num} to {new_num} replicas. "
+                f"{curr_stats_str}"
```
[the remark this answers is not in the document]
Comment (edoakes) [reply python/ray/serve/_private/deployment_state.py:1688]: upscaling/downscaling matches our states

File: python/ray/serve/_private/deployment_state.py (hunk 2 above)
Comment (edoakes) [reply python/ray/serve/_private/deployment_state.py]: done

File: python/ray/serve/_private/deployment_state.py [hunk 6]
```diff
@@ -1608,6 +1609,27 @@ def get_total_num_requests(self) -> float:
...
+                    del self.handle_requests[handle_id]
+                    total_ongoing_requests = handle_metric.queued_requests + sum(
+                        handle_metric.running_requests.values()
+                    )
+                    if total_ongoing_requests > 0:
+                        logger.info(
+                            f"Dropping stale metrics for handle '{handle_id}' "
+                            f"because no update was received for {timeout_s:.1f}s. "
+                            f"Ongoing requests was: "
+                            "{total_ongoing_requests}."
```
Comment (zcin) [inline python/ray/serve/_private/deployment_state.py:1630]:
```suggestion
                            f"Ongoing requests was: {total_ongoing_requests}."
```

Comment (edoakes) [reply python/ray/serve/_private/deployment_state.py:1630]: what the heck the linter sabotaged me here!

**Pull request merged**